library/general/src/lib/ui/multi_messages_dialog.rb

```diff
@@ -0,0 +1,247 @@
+      move_to(@position - 1)
+      true
+    end
+
+    # 'Close' button handler
+    #
+    # Closes the dialog returning :close
+    def close_handler
+      finish_dialog(:close)
+    end
+
+    # Display the message in the given position
+    #
+    # @param position [Integer] Position of the message to be considered as
+    #                           'current'.
+    def move_to(position)
+      @position = position
+      refresh
+    end
+
+    # Timeout handler
+    #
+    # Handles timeouts generated by TimeoutUserInput updating the
+    # remaining_time counter and the timer. If no time left, it closes
+    # the dialog.
+    def timeout_handler
+      @remaining_time -= 1
+      if @remaining_time.zero?
+        finish_dialog(:timeout)
+      else
+        Yast::UI.ChangeWidget(Id(:timer), :Value, @remaining_time.to_s)
+      end
+    end
+
+    # Stop handler
+    #
+    # Disable timeout counter
+    def stop_handler
+      stop_timer
+      Yast::UI.ChangeWidget(Id(:stop), :Enabled, false)
+    end
+
+    # Stop timer
+    #
+    # Stop the timer countdown.
+    def stop_timer
+      @timeout = false
+    end
+
+    # User input handling
+    #
+    # If the dialog is timed, it relies Yast::UI::TimeoutUserInput.
+    # Otherwise, relies on Yast::UI::UserInput
+    #
+    # @return [Symbol] User input or :timeout symbol
+    def user_input
+      if timed?
+        Yast::UI::TimeoutUserInput(TIMEOUT_STEP)
+      else
+        super
+      end
+    end
+
+    # Determines whether the dialog is timed or not
+    #
+    # @return [true,false] True if the dialog is timed; false otherwise.
+    def timed?
+      timeout.is_a?(Integer)
+    end
+
+  private
+
+    # Dialog options
+    #
+    # Overrides UI::Dialog#dialog_options
+    #
+    # @return [Yast::Term] Dialog options.
+    #
+    # @see UI::Dialog#dialog_options
+    def dialog_options
+      Opt(:decorated)
+    end
+
+    # Refresh message and buttons
+    #
+    # Update the message and enables/disables 'Next' and 'Back' buttons
+    # accordingly.
+    #
+    # @see #current_message
+    # @see #first_message?
+    # @see #last_message?
+    def refresh
+      Yast::UI.ReplaceWidget(Id(:body), message_body_ui)
+      Yast::UI.ChangeWidget(Id(:back), :Enabled, !first_message?)
+      Yast::UI.ChangeWidget(Id(:next), :Enabled, !last_message?)
+    end
+
+    # Returns the current message
+    #
+    # That's the message that it's supposed to be shown to the user at a given
+    # time.
+    #
+    # @return Object
+    def current_message
+      messages[@position]
+    end
+
+    # Returns the amount of messages
+    #
+    # It is just a convenience method.
+    #
+    # @return [Integer] Amount of messages.
+    def total
+      messages.size
+    end
+
+    # UI content describing the UI for the current message's body
+    #
+    # @return [Yast::Term] UI content for dialog
+    def message_body_ui
+      title =
+        if messages.size > 1
```

Comment:
NP: Um, I don't like much `if` condition in assignment, it's quite unusual in the YaST code. I'd rather prefer the ternary operator here.